kOS/pegas.ks

```diff
@@ -71,56 +71,105 @@ setComms(). 			//	Setting up communications
+
+//initial data for bilinear control of pitch, if used.
+IF controls:HASKEY("pitchControl") {
+	SET pitchCtrlTrigger TO controls["pitchStartAlt"].
+	SET pitchCtrlIterator TO controls["pitchControl"]:iterator.
+}
+
 //	Main loop - wait on launch pad, lift-off and passive guidance
 UNTIL ABORT {
 	//	Sequence handling
 	IF systemEventFlag = TRUE { systemEventHandler(). }
 	IF   userEventFlag = TRUE {   userEventHandler(). }
 	IF  commsEventFlag = TRUE {  commsEventHandler(). }
-	//	Control handling
-	IF ascentFlag = 0 {
-		//	The vehicle is going straight up for given amount of time
-		IF TIME:SECONDS >= liftoffTime:SECONDS + controls["verticalAscentTime"] {
-			//	Then it changes attitude for an initial pitchover "kick"
-			SET steeringVector TO aimAndRoll(HEADING(mission["launchAzimuth"],90-controls["pitchOverAngle"]):VECTOR, steeringRoll).
-			SET ascentFlag TO 1.
-			pushUIMessage( "Pitching over by " + ROUND(controls["pitchOverAngle"],1) + " degrees." ).
+
+	// linear control of pitch angle against Altitude is used by default if "pitchControl" key exists:
+	IF controls:HASKEY("pitchControl") {
+
+			//Resetting the a and b values of the linear regression on each alitude trigger
+			IF SHIP:ALTITUDE > pitchCtrlTrigger {
+				pitchCtrlIterator:NEXT.
+
+				//Specific to the first ttime.
+				IF pitchCtrlIterator:INDEX = 0 {
+					set beginPitch to 90.
+					set beginAlt to controls["pitchStartAlt"].
+				}
+
+				ELSE{
+					set beginPitch to controls["pitchControl"][pitchCtrlIterator:INDEX-1]["endPitch"].
+					set beginAlt to controls["pitchControl"][pitchCtrlIterator:INDEX-1]["endAlt"].
+				}
+
+				set endPitch to controls["pitchControl"][pitchCtrlIterator:INDEX]["endPitch"].
+				set endAlt to controls["pitchControl"][pitchCtrlIterator:INDEX]["endAlt"].
+
+				set aPitch to (endPitch-beginPitch)/(endAlt-beginAlt).
+				set bPitch to  endPitch-(aPitch*endAlt).
+
+
+				SET pitchCtrlTrigger to controls["pitchControl"][pitchCtrlIterator:INDEX]["endAlt"].
+				pushUIMessage( "Pitch follow " + toSCI(aPitch) + " * ALT + " + toSCI(bPitch)).
+			}
+
+		IF DEFINED aPitch and DEFINED bPitch {
+			SET steeringVector TO aimAndRoll(HEADING(mission["launchAzimuth"],aPitch*SHIP:ALTITUDE+bPitch):VECTOR, steeringRoll).
 		}
 	}
-	ELSE IF ascentFlag = 1 {
-		//	It keeps this attitude until velocity vector matches it closely
-		IF TIME:SECONDS < liftoffTime:SECONDS + controls["verticalAscentTime"] + 3 {
-			//	Delay this check for the first few seconds to allow the vehicle to pitch away from current prograde
-		} ELSE {
-			//	Attitude must be recalculated at every iteration though
-			SET velocityAngle TO VANG(SHIP:UP:VECTOR, SHIP:VELOCITY:SURFACE).
-			IF controls["pitchOverAngle"] - velocityAngle < 0.1 {
+
+
+	//If no "pitchControl" key exists, use the prograde tracking mode:
+	ELSE{
+		SET ascentFlag TO 0.	//	0 = vertical, 1 = pitching over, 2 = notify about holding prograde, 3 = just hold prograde
```

Comment:
This line should stay outside of the loop. By putting it here, you effectively set the ascentFlag to 0 at the beginning of every iteration so it will never advance past 0, breaking the original ascent mode.

Reply:
Well I havent' see that. There is also another bug with my code that I need to solve:
If you finish your `pitchControl` before the start of UPFG guidance, the program crashes because my iterator still increases and try to find a non existing index in the `pitchControl` LIST. Il will correct that.

Reply:
I'll try this ascent mode out with my other rockets and see if I can spot any other issues during normal usage.

As for that issue with `pitchControl` finishing before start of UPFG, I guess one option would be to fallback to following prograde after that point, maybe interpolate between current pitch and prograde or something like that?

Reply:
Hi, I will Think a little more, make some tests, and send you another pull request when I have something that seems correct.
I also have others ideas for PEGAS, maybe we can discuss of it one time.